core/src/main/scala/za/co/absa/spline/core/transformations/LineageProjectionMerger.scala

```diff
@@ -60,7 +60,41 @@ object LineageProjectionMerger extends AsyncTransformation[DataLineage] {
     }
 
     val dataTypes = {
-      val dataTypeIds = attributes.map(_.dataTypeId) // todo: ++ expressions' data types IDs
+      val expressions = operations.flatMap {
+        case op: ExpressionAware => op.expressions
+        case _ => Nil
+      }
+
+      val expressionTypeIds: Set[UUID] = {
+        def traverseAndCollect(accumulator: Set[UUID], expressions: List[Expression]): Set[UUID] = expressions match {
+          case Nil => accumulator
+          case exp :: queue =>
+            val updatedAccumulator = exp match {
+              case tex: TypedExpression => accumulator + tex.dataTypeId
+              case _ => accumulator
+            }
+            val updatedQueue = exp.children.toList ++ queue
+            traverseAndCollect(updatedAccumulator, updatedQueue)
+        }
+
+        traverseAndCollect(Set.empty, expressions.toList)
+      }
+
+      val dataTypeIds = {
```

Comment:
This variable name could be more descriptive e.g. usedDataTypes or retainedDataTypes.

Reply:
renamed to `retainedDataTypes`